fix: remove memory leak in stack

vil02 opened the pull request:
#### Description of Change
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Contributors guide: https://github.com/TheAlgorithms/C-Plus-Plus/CONTRIBUTING.md
-->
I added the member function `deleteAllNodes()` into the `stack` and applied it in the destructor, assignment operator and in the `clear` method. I have applied address sanitizer and valgrind on [`test_stack.cpp`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/97c7d918789ce1c414e1a7737b766ceac0b4f6db/data_structures/test_stack.cpp) and [`test_stack_students.cpp`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/97c7d918789ce1c414e1a7737b766ceac0b4f6db/data_structures/test_stack_students.cpp) - no further leaks were found. The mentioned files, are the only ones using `stack.h`.

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] Added description of change
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: <!-- Please add a one-line description for developers or pull request viewers -->
Quickfix of the memory leak in [`stack.h`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/97c7d918789ce1c414e1a7737b766ceac0b4f6db/data_structures/stack.h).

martin-sos commented on data_structures/stack.h:
```diff
@@ -73,8 +69,20 @@ class stack {
     }
 
+ private:
+    void deleteAllNodes() {
```
would it be an improvement to use stackTop instead of curNode to maintain the top of the stack? 

```suggestion
void deleteAllNodes() {
   while (stackTop != nullptr) {
            auto tmpNode = stackTop;
            stackTop = stackTop->next;
            delete tmpNode;
        }
}
```

vil02 replied:
You are absolutely right! Thank you for this remark.

martin-sos commented on data_structures/stack.h:
```diff
@@ -109,16 +115,16 @@ class stack {
-        if (stackTop != nullptr) {
-            stackTop = nullptr;
-        }
+        deleteAllNodes();
```
from here on until line 147 the size of _this_ is not accurate anymore. What about resetting the size within deleteAllNodes?

martin-sos replied:
in this case, line 120 in clear() would not be needed anymore, as well.

vil02 replied:
I do not see the variable `size` to be _really_ used anywhere (well the method `display()` reads it, note that the `size` could be computed there; there is nothing like `get_size()` - having it would justify presence of the `size` member). I think that this is beyond the scope of this PR.

martin-sos replied:
Yes, I'm fine with that.

martin-sos reviewed (commented):
As top(), display() and isEmptyStack() are non modifying methods, you can qualify them with const.

vil02 commented:
> As top(), display() and isEmptyStack() are non modifying methods, you can qualify them with const.

Agree - this implementation needs some _cleaning_ - making it _const correct_ and really _testable_ is something what should be done quickly. But I think this is beyond the scope if this PR.

vil02 commented:
> To remove a few [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/actions/runs/3245717513/jobs/5333516869) errors, please change the filename to `stack.hpp`. You might need to update other instances as well.

This change indeed _suppressed_ the errors. But as discussed above: this implementation needs a cleanup.

The entire [`Awesome CI Workflow `](https://github.com/vil02/C-Plus-Plus/actions/runs/3250645588) run on my side. @Panquesito7 could you please approve the run on _your end_?

Panquesito7 commented on data_structures/stack.hpp:
```diff
@@ -1,10 +1,10 @@
 /**
- * @file stack.h
+ * @file stack.hpp
```
No need to add the filename. It's best to let Doxygen automatically detect it, otherwise, you'd have to specify the whole directory as well.
```suggestion
 * @file
```

Panquesito7 commented:
Would you like to work on these files and improve them to our standards? That would fit for another PR, of course.

Panquesito7 reviewed (approved):
Great work! Thanks. 🚀

vil02 commented:
> Would you like to work on these files and improve them to our standards? That would fit for another PR, of course.

I could do that. However there are few things, which I would like to clarify before:
- tests: should I use some framework for writing them (something like [GoogleTest](https://github.com/google/googletest) or [Boost.Test](https://www.boost.org/doc/libs/1_80_0/libs/test/doc/html/index.html)). I realize that this is a bigger topic. 
- I think for the _educational purposes_ it might be a good idea to stay with _raw pointers_ instead of using _smart pointer_. Do you agree?
- there is a similar memory leak in [`queue.h`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/59a0d3a06906140d95d53ba783d66ec11dd3afc1/data_structures/queue.h).

Panquesito7 commented:
> > Would you like to work on these files and improve them to our standards? That would fit for another PR, of course.
> 
> I could do that. However there are few things, which I would like to clarify before:
> 
> * tests: should I use some framework for writing them (something like [GoogleTest](https://github.com/google/googletest) or [Boost.Test](https://www.boost.org/doc/libs/1_80_0/libs/test/doc/html/index.html)). I realize that this is a bigger topic.
> * I think for the _educational purposes_ it might be a good idea to stay with _raw pointers_ instead of using _smart pointer_. Do you agree?
> * there is a similar memory leak in [`queue.h`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/59a0d3a06906140d95d53ba783d66ec11dd3afc1/data_structures/queue.h).

1. I haven't heard of that tool, but if it makes good self-test implementations @tjgurwara99, thoughts?
2. Smart pointers are used in modern C++ which we should stick to, IMHO. This will also help others learn smart pointers and make the CI happy as well.
3. Great finding, thank you! Feel free to fix that if you like in another PR. :)

vil02 commented:
@Panquesito7 I changed the implementation to use `std::shared_ptr` and added [`test_stack.cpp`](https://github.com/vil02/C-Plus-Plus/blob/5167744a2380687a5c501e08dc074c6ff23f76de/data_structures/test_stack.cpp). The old tests are now in [`test_stack_legacy.cpp`](https://github.com/vil02/C-Plus-Plus/blob/5167744a2380687a5c501e08dc074c6ff23f76de/data_structures/test_stack_legacy.cpp).

vil02 commented:
I think now it everything is ready to be merged.

Panquesito7 commented on data_structures/test_stack_legacy.cpp:
```diff
@@ -0,0 +1,61 @@
+#include <iostream>
```
We should probably specify this is using raw pointers and not smart pointers probably?

vil02 replied:
This file is the file with the _old tests_. It uses the same `Stack` as the other tests (i.e. currently with `std::shared_ptr`). In my opinion it could be removed - do you agree on that?

Panquesito7 replied:
Unless it is for some sort of educational purpose, it can be removed. 🙂

vil02 replied:
For educational reasons, I think it is better to remove it. Thanks to that, I can change make `top` and `pop` throw an [`std::invalid_argument`](https://en.cppreference.com/w/cpp/error/invalid_argument) when the stack is empty.

tjgurwara99 commented on data_structures/test_stack.cpp:
```diff
@@ -1,59 +1,214 @@
-#include "./stack.h"
+#include "./stack.hpp"
+
+void assertTrue(const bool inValue) {
+    if (!inValue) {
+        std::cout << "Assertion fails!\n";
+        exit(1);
+    }
+}
```
I don't understand the reasoning behind creating this when `assert` is already included in the file? `assert(a == b)` does the same thing so I don't understand the use case here, unless you're worried about `NDEBUG` which shouldn't be a concern here since we don't use that env variable (from what I can tell).

vil02 replied:
I am concern about the `NDEBUG` and the [_standard_ `assert`](https://en.cppreference.com/w/cpp/error/assert). Tests should be able to fail no matter what kind of build is used and how the build is configured.

As [mentioned earlier](https://github.com/TheAlgorithms/C-Plus-Plus/pull/2235#issuecomment-1279699322): the perfect solution requires some proper testing framework.

tjgurwara99 replied:
Your concern for the `NDEBUG` is valid but unnecessary I think - maybe @Panquesito7 can shed some light on this. The reason why I think it's unnecessary is that these `assert` statements are enough by themselves within this repositories build context. You say the tests should fail in all build scenarios but when `NDEBUG` is enabled the tests are not tests anymore and this repository would never add the `NDEBUG` macro to begin with so that shouldn't be something you should be concerned about.

vil02 replied:
@tjgurwara99 based on the :+1: of @Panquesito7 I think you insist on using the standard assert. Now this is done. Just for information: the [CI](https://github.com/vil02/C-Plus-Plus/actions/runs/3856762403/jobs/6573363833) passes on my end.

vil02 commented on data_structures/stack.hpp:
```diff
@@ -0,0 +1,108 @@
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <memory>
+#include <stdexcept>
```
Done.

Panquesito7 reviewed (approved):
Amazing. Thank you! 🚀
